### PR TITLE
[feature] Typed extra for heading

### PR DESCRIPTION
**Pull request by llambeau**
+ Add acceptance test for tuple type with extra
+ Add acceptance test for tuple type with typed extra

### CODE REVIEW

File: features/finitio/dressing/relation-type-with-extra.feature [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    Given I dress JSON's '[{ "name": "Finitio", "age": 1 }]' with Info
+    Then the result should be a representation for Info
+
+  Scenario: Dressing when an invalid tuple
+
+    Given I dress JSON's '[{ "name": "Finitio" }]' with Info
+    Then it should be a TypeError
+    And its root cause should be:
+      | message                 |
+      | Missing attribute `age` |
```
Comment (blambeau) [inline features/finitio/dressing/relation-type-with-extra.feature]: Why so??

File: features/finitio/dressing/relation-type-with-extra.feature [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+    Given I dress the following JSON document with Info:
+      """
+      [{ "name": "Finitio", "age": 1 },
+       { "name": "Finitio", "age": 1 }]
+      """
+    Then it should be a TypeError
+    And its root cause should be:
+      | message                                       |
+      | Duplicate Tuple: `{"name":"Finitio","age":1}` |
```
Comment (blambeau) [inline features/finitio/dressing/relation-type-with-extra.feature]: I'm not 100% sure we want to test this here... well, maybe we want because of implementation details but the "no duplicates" feature seems orthogonal.

File: features/finitio/dressing/tuple-type-with-typed-extra.feature [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    Given the System is
+      """
+      Age   = Integer( i | i>=0 )
+      Info  = { name : String, ...: Integer }
+      """
+
+  Scenario: Dressing a valid tuple with extra
+
+    Given I dress JSON's '{ "name": "Finitio", "age": 1 }' with Info
+    Then the result should be a representation for Info
```
Comment (blambeau) [inline features/finitio/dressing/tuple-type-with-typed-extra.feature]: Please add another Integer-typed extra entry to the json input

File: features/step_definitions/steps.coffee [hunk 4]
```diff
@@ -29,7 +29,10 @@ module.exports = ->
       callback()
     catch e
       error = e
-      console.log(e.explainTree())
+      if (e.explainTree)
+        console.log(e.explainTree())
+      else
+          console.log(e.message)
```
Comment (blambeau) [inline features/step_definitions/steps.coffee]: spacing issue: unindent

File: features/step_definitions/steps.coffee [hunk 5]
```diff
@@ -108,11 +111,20 @@ module.exports = ->
       should(type.heading.getAttr(a1).required).be.true
     callback()
 
-  @Then /^it allows extra attributes$/, (callback) ->
+  @Then /^it allows extra attributes$/, (name, callback) ->
```
Comment (blambeau) [inline features/step_definitions/steps.coffee]: Remove `name` here

File: src/finitio/support/heading.coffee [hunk 6]
```diff
@@ -14,7 +15,7 @@ class Heading
   Fetchable  this, 'attributes', 'attribute', (name)-> this.getAttr(name)
 
   DEFAULT_OPTIONS = {
-    allowExtra: false
+    allowExtra: null
```
Comment (blambeau) [inline src/finitio/support/heading.coffee]: Keep false.

File: src/finitio/support/meta.js [hunk 7]
```diff
@@ -171,10 +171,18 @@ module.exports = (function(){
...
+        Attribute.info({ name: 'type', type: Meta.Type })
+      ]
+    })
+  });
+
   Meta.HeadingOptions = TupleType.info({
     heading: Heading.info({
       attributes: [
-        Attribute.info({ name: 'allowExtra', type: Js.Boolean })
+        Attribute.info({ name: 'allowExtra', type: Meta.HeadingAllowExtraOption })
```
Comment (blambeau) [inline src/finitio/support/meta.js]: Broken API on generated .json in the past, please keep accepting booleans.

File: src/finitio/support/heading.coffee [hunk 8]
```diff
@@ -45,11 +46,21 @@ class Heading
...
-    @options['allowExtra']
+  allowExtra: (value) ->
+    unless @options.allowExtra?
+      return false
+
+    unless value?
+      return true
+
+    extraType = @options.allowExtra
+    extraType.include(value)
```
Comment (blambeau) [inline src/finitio/support/heading.coffee]: Keep `allowExtra` with a type/subtype semantics regarding the parameter. Add `allowExtraValue` instead.

File: src/finitio/support/heading.coffee [hunk 9]
```diff
@@ -45,11 +46,21 @@ class Heading
...
+
+  getExtraType: ->
+    @options.allowExtra
 
   multi: ->
-    @options['allowExtra'] || $u.any(@attributes, (a) -> !a.required)
+    @options['allowExtra']? || $u.any(@attributes, (a) -> !a.required)
```
Comment (blambeau) [inline src/finitio/support/heading.coffee]: Not sure it is still necessary if we remove `null` above.

File: src/finitio/support/heading.coffee [hunk 10]
```diff
@@ -71,7 +84,13 @@ class Heading
...
     $u.every(l, (a)-> not(a.required)) and
-    (@allowExtra() or not(other.allowExtra())) and
+    (
+      not(other.allowExtra()) or
+      (
+        @allowExtra() and
+        @options.allowExtra._isSuperTypeOf(other.options.allowExtra)
+      )
+    )
     (@allowExtra() or $u.isEmpty(r))
```
Comment (blambeau) [inline src/finitio/support/heading.coffee]: At least a `and` is missing

File: src/finitio/type/tuple_type.coffee [hunk 11]
```diff
@@ -47,6 +47,15 @@ class TupleType extends Type
...
           m
+      else if !attr? and @heading.allowExtra()
+        type = @heading.getExtraType()
+        subm = type.mDress(attrValue, Monad)
+        subm.onFailure (error)->
+          error.location = attrName
+          subm
+        subm.onSuccess (val)->
+          result[attrName] = val
+          success
```
Comment (blambeau) [inline src/finitio/type/tuple_type.coffee]: Some cases later in the if/then/else might end up being unnecessary and must be removed/cleaned.

**Comment by llambeau** [comment]
@blambeau ready for CR. @JosephJNK if you wanna have a look

**Comment by blambeau** [comment]
@llambeau mostly OK. I would just keep a the `expect(...).equal(...)` fix in a separate commit. You do it, or I do it (but I'll then do it later).

**Comment by blambeau** [comment]
@llambeau Excellent work! great feature.
